Refs #58 - Allow for typed NULLs in Java calls

freakboy3742 opened the pull request:
Many Java APIs allow the user to pass `null` as an argument. However, as we need to do dynamic lookup of methods by type, a `None` argument can't be easily matches to a specific type.

This change allows the user to pass in `None` as an argument to methods where an object is expected. 

If there are multiple polymorphic forms that *could* accept None, an error is raised. There is an edge case where this is legal - a class with `mymethod(BaseClass x)` and `mymethod(SubClass x)` will be use the SubClass version if passed a `None` . This case raises an error at present.

The implementation also changes the key used to look up polymorphs. Previously it was the JNI signature; we now use a tuple composed of the signature's parts. This should be no slower for lookup; but allows us to do wildcard matching on the JNI signature parts.

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

t-arn commented:
@freakboy3742 I like it, that we now can simply pass in None for null and rubicon will try to find the correct match. This is like it works on Java what makes it intuitive to use. 

Your original idea of introducing a typed NULL would have avoided the seldom cases where we have several matches, but it would have been less intuitive to use.

freakboy3742 commented:
@t-arn As it happens, I actually prototyped the typed NULL first, and reworked the solution to what I've implemented here. However, I'm not 100% convinced that was the right choice. 

In practice, what I've implemented in this PR would behave like this:
```
result = call_some_method(42, None)
if result is None:
   call_other_method(result)
```
This is more "Pythonic" - it uses Python objects, and transparently resolves how that maps to Java. However, the lookup process is more complex (and therefore potentially slower). There's also the edge case around ambiguous method prototypes that I mentioned.

The alternative is typed nulls:
```
from rubicon.java import NULL

result = call_some_method(42, MyObject.__null__)
if result == NULL:
    call_other_method(result)
```
This has a much simpler implementation. It's also *faster* (because you do a specific look up for the prototype you want), and it doesn't have the ambiguity problem (again, because you're doing a specific lookup). The downside is that it's less "Pythonic", because you need to use Java objects not Python syntax. This code "tastes" more like Java (e.g., you do an equality with NULL, rather than `is None`). In the Toga context, any code that invokes a Java API would need to be wrapped to check for the `== NULL` condition and converted into a Python `None`.

However, this API is also a language bridging layer, so a certain amount of Java flavour should be expected :-) and by being "smart" with the None casting, our implementation becomes less explicit. Java is, after all, a typed language, so we shouldn't be surprised when our usage of Java requires typing.

Do you have any thoughts/preferences in those two options?

t-arn commented:
I wonder how the typed NULL approach would work for Interfaces. Would we always need to provide an implementation, so we could create an object that we could then set to NULL?

freakboy3742 commented:
The process for Interfaces and Typed nulls wouldn't be too far removed from the process for classes - it would be something like:

```
MyInterface = JavaInterface('com/example/MyInterface')

some_method(MyInterface.__null__)
```

Internally, all we're doing is defining a placeholder object that can identify itself as "Lcom/example/MyInterface;" when building the JNI prototype for the method we're looking for.

t-arn commented:
So, the typed NULL approach is also pretty easy to use and it does not have the potential ambiguity problem. Still, without some documentation, nobody will know how to use it. Therefore, my vote is the None approach.

freakboy3742 commented:
@t-arn Ok - but that's an argument for "write documentation" not "pick an option on the basis that documentation doesn't exist". Assuming there was better (or any!) documentation - does that change your opinion?

t-arn commented:
Well, yes it actually does. The typed NULL *has* technical advantages: No ambiguity problem, faster, less complex code.
How about this: In addition to properly document the usage we could provide a specific error message for the case where None is used as a parameter hinting to the typed NULL documentation?

t-arn commented:
@paulproteus What implementation do you prefer? The None implementation of this PR or the typed NULL implementation?

t-arn commented:
@freakboy3742 Hello Russel, do you intend to change the implementation of this PR and use the typed NULL approach? I need to know because I want to implement the ProgressBar for Android which needs this PR. And I don't want to use the current implementation and then be forced to change my code later on.
Tom

freakboy3742 commented:
@t-arn I think I'm leaning towards the typed NULL approach. It's a simpler implementation, and more explicit; plus, it doesn't preclude introducing implicit NULLs at some point in the future. 

Hopefully I'll get some time this weekend to wrap up this PR.
